crucible-mir/src/Mir/Intrinsics.hs

```diff
@@ -588,6 +590,108 @@ muxRef sym iTypes c (MirReferenceMux mt1) (MirReferenceMux mt2) =
+ where MirAggregateRepr = IntrinsicRepr (knownSymbol @MirAggregateSymbol) Empty
+
+type family MirAggregateFam (sym :: Type) (ctx :: Ctx CrucibleType) :: Type where
+  MirAggregateFam sym EmptyCtx = MirAggregate sym
+  MirAggregateFam sym ctx = TypeError
+    ('Text "MirAggregateType expects no arguments, but was given" ':<>: 'ShowType ctx)
+
+instance IsSymInterface sym => IntrinsicClass sym MirAggregateSymbol where
+  type Intrinsic sym MirAggregateSymbol ctx = MirAggregateFam sym ctx
+
+  muxIntrinsic sym tys _nm Empty = muxMirAggregate sym tys
+  muxIntrinsic _sym _tys nm ctx = typeError nm ctx
+
+muxMirAggregateEntry :: forall sym.
+  IsSymInterface sym =>
+  sym ->
+  IntrinsicTypes sym ->
+  Word ->
+  Pred sym ->
+  Maybe (MirAggregateEntry sym) ->
+  Maybe (MirAggregateEntry sym) ->
+  IO (MirAggregateEntry sym)
+muxMirAggregateEntry sym itefns offset c mEntry1 mEntry2 =
+  case (mEntry1, mEntry2) of
+    (Nothing, Nothing) -> error "impossible"
+    (Just entry1, Nothing) -> goOneSided c entry1
+    (Nothing, Just entry2) -> do
+      c' <- notPred sym c
+      goOneSided c' entry2
+    (Just entry1, Just entry2) -> goTwoSided entry1 entry2
+  where
+    goOneSided :: Pred sym -> MirAggregateEntry sym -> IO (MirAggregateEntry sym)
+    goOneSided activeCond (MirAggregateEntry sz tpr rv) = do
+      rv' <- muxRegForType sym itefns (MaybeRepr tpr) activeCond rv Unassigned
+      return $ MirAggregateEntry sz tpr rv'
+
+    goTwoSided :: MirAggregateEntry sym -> MirAggregateEntry sym -> IO (MirAggregateEntry sym)
+    goTwoSided (MirAggregateEntry sz tpr rv1) (MirAggregateEntry sz' tpr' rv2) = do
+      when (sz' /= sz) $
+        die $ "entry size " ++ show sz ++ " != " ++ show sz'
+      Refl <- case testEquality tpr tpr' of
+        Just x -> return x
+        Nothing -> die $ "type mismatch: " ++ show tpr ++ " != " ++ show tpr'
+      rv' <- muxRegForType sym itefns (MaybeRepr tpr) c rv1 rv2
+      return $ MirAggregateEntry sz tpr rv'
+
+    die :: String -> IO a
+    die msg = fail $ "bad MirAggregate merge: offset " ++ show offset ++ ": " ++ msg
+
+muxMirAggregate :: forall sym.
+  IsSymInterface sym =>
+  sym ->
+  IntrinsicTypes sym ->
+  Pred sym ->
+  MirAggregate sym ->
+  MirAggregate sym ->
+  IO (MirAggregate sym)
+muxMirAggregate sym itefns c (MirAggregate sz1 m1) (MirAggregate sz2 m2) = do
+  when (sz1 /= sz2) $ do
+    fail $ "bad MirAggregate merge: size " ++ show sz1 ++ " != " ++ show sz2
+  m' <- sequence $ IntMap.mergeWithKey
+    (\offset entry1 entry2 -> Just $ muxEntries offset (Just entry1) (Just entry2))
+    (IntMap.mapWithKey $ \offset entry1 -> muxEntries offset (Just entry1) Nothing)
+    (IntMap.mapWithKey $ \offset entry2 -> muxEntries offset Nothing (Just entry2))
+    m1 m2
+  return $ MirAggregate sz1 m'
+  where
+    muxEntries off e1 e2 = muxMirAggregateEntry sym itefns off' c e1 e2
+      where off' = (fromIntegral :: Int -> Word) off
```

Comment:
Style nit pick: I'd prefer to use type applications here instead:

```hs
      where off' = fromIntegral @Int @Word off
```

Reply:
I think the type ascription is clearer because the reader doesn't need to know which type parameter is which

Reply:
OK. If you go that route, I would personally prefer to just give `off'` a top-level type signature (e.g., `off' :: Word; off' = ...`), since that adds less visual clutter in the expression part. I don't feel strongly about this, however, so I'll leave this to your judgment.

Reply:
I removed the ascription entirely. It was necessary for type inference at some point, but I guess not any more.